README.md
 # starreviews
 
A flutter package for displaying detailed star reviews

<img src=https://github.com/buraktabn/starreviews/raw/master/images/example.png width="200" />

## Install

```yaml
dependencies:  
  starreviews: ^1.0.4
```

```console
$  flutter pub get
```

```dart
import  'package:starreviews/starreviews.dart';
```

## Usage

```dart
List<String> _names = ['Star 5', 'Star 4', 'Star 3', 'Star 2', 'Star 1'];  
List<double> _values = [0.5, 0.2, 0.8, 0.5, 0.1];
```
```dart
StarReviews(  
    total: 20,  
    starNames: _names,  
    values: _values,  
    showPercentage: true,  
    average: 3.2,
)
```
```dart
StarReviewsHorizontal(
    total: 30,
    starNames: _names,
    showProgressBarBorder: false,
    valueColor: Colors.black,
    progressBarBackgroundColor: Colors.grey.withOpacity(0.4),
    values: _values,
    showPercentage: true,
    starColor: Colors.black,
    average: 4.2,
)
```

## Customization
Here is a list of properties available to customize the widget:

|        Name        	|       Type      	|                 Description                	|
|:------------------:	|:---------------:	|:------------------------------------------:	|
| total| int| total numbers of reviews           	|
| showHeader| bool| if false, hides header           	|
| showBottom| bool| if false, hides bottom reviews           	|
| showPercentage| bool| if false, hides rating number           	|
| starNames| List<String>| list of star names (max 5)           	|
| percentageStyle| TextStyle| style applied to percentage                                  	|
| showPercentage| bool| if false, hides percentage                                 	|
| valueColor| Color| color of the progressbar             	|
| progressBarBackgroundColor| Color| color of the progress bar background                      	|
| values| List<double>| list of review values as `double`                     	|
| starSize| double| the size of the stars                     	|
| starColor| Color| the color of the stars                     	|
| average| double| the average number that'll be displayed                     	|
| showProgressBarBorder| bool| option to hide borders of the progress bar                     	|
| lineHeight| double| height of the progress bar                     	|
| spaceBetween| double| the space between progress bars                     	|

  
## Contributions

Contributions of any kind are more than welcome! Feel free to fork and improve in any way you want, make a pull request, or open an issue.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## 1.0.0 - 2020-05-02

### Added

-   Initial Commit

## 1.0.1 - 2020-05-02

### Changed

-   Changed average star display

## 1.0.2 - 2020-06-22

### Changed

-   Added dynamic number feature
-   Added a new design `StarReviewsHorizontal`

## 1.0.3 - 2020-06-22

### Changed

-   Added new params for customizable

## 1.0.4 - 2020-06-22

### Changed

-   Added a new param for customizable

## 1.0.5 - 2020-07-23

### Changed

-   Fixed a bug



test/starreviews_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:starreviews/starreviews.dart';

void main() {
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:buraktaban/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/progressbar.dart
import 'package:flutter/material.dart';
import 'package:percent_indicator/linear_percent_indicator.dart';

class ProgressBar extends StatelessWidget {
  final Color valueColor;
  final Color backgroundColor;
  final double value;

  ProgressBar({Key key, this.valueColor, this.backgroundColor, this.value})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Container(
      padding: EdgeInsets.all(2),
      decoration: BoxDecoration(
          borderRadius: BorderRadius.all(Radius.circular(8)),
          border: Border.all(width: 1, color: const Color(0xffcecece))),
      child: ClipRRect(
        borderRadius: BorderRadius.circular(8),
        child: Container(
          height: 10,
          child: LinearPercentIndicator(
            lineHeight: 10,
            padding: null,
            linearStrokeCap: LinearStrokeCap.roundAll,
            percent: this.value,
            // percent filled
            progressColor: this.valueColor,
            backgroundColor: this.backgroundColor,
          ),
        ),
      ),
    );
  }
}


lib/starreviews.dart
library starreviews;

import 'dart:math';

import 'package:flutter/material.dart';
import 'package:starreviews/progress_bar/progress_bar_layout.dart';
import 'package:starreviews/star_display.dart';

class StarReviews extends StatelessWidget {
  final int total;
  final bool showHeader;
  final bool showBottom;
  final bool showRatingNumber;

  final List<String> starNames;
  final TextStyle starNameStyle;
  final TextStyle percentageStyle;
  final bool showPercentage;

  final Color valueColor;
  final Color progressBarBackgroundColor;
  final List<double> values;

  final double starSize;
  final Color starColor;

  final double average;
  final bool showProgressBarBorder;

  final double lineHeight;
  final double spaceBetween;

  StarReviews(
      {Key key,
      @required this.total,
      @required this.starNames,
      this.starSize = 16,
      this.lineHeight = 10,
      this.spaceBetween = 11,
      this.showProgressBarBorder = true,
      this.starColor = const Color(0xffffd900),
      this.showRatingNumber = true,
      this.showHeader = true,
      this.showBottom = true,
      this.showPercentage = true,
      this.starNameStyle = const TextStyle(fontSize: 12),
      this.percentageStyle = const TextStyle(fontSize: 12),
      this.valueColor = const Color(0xff656565),
      this.progressBarBackgroundColor = Colors.white,
      @required this.average,
      @required this.values})
      : assert(total != null),
        super(key: key) {
    if (values == null) {
      throw ArgumentError('value cannot be empty');
    }

    if (starNames.length != values.length) {
      throw ArgumentError('startNames and values\' length different');
    }

    if (starNames == null) {
      throw ArgumentError('starNames cannot be empty');
    }
  }

  @override
  Widget build(BuildContext context) {
    return Container(
        child: Column(
      children: <Widget>[
        Visibility(
          visible: this.showHeader,
          child: Container(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: <Widget>[
                Row(
                  children: <Widget>[
                    StarDisplay(
                      value: this.average,
                      size: starSize,
                      starCounts: this.starNames.length,
                    ),
                    SizedBox(
                      width: 5,
                    ),
                    Text(
                      '${this.average} out of ${this.starNames.length}',
                      style: TextStyle(fontSize: 14),
                    )
                  ],
                ),
                Visibility(
                  visible: showRatingNumber,
                  child: Column(
                    children: <Widget>[
                      SizedBox(
                        height: 5,
                      ),
                      Text(
                        '${this.total} ratings',
                        style: TextStyle(
                            color: const Color(0xff919191), fontSize: 12),
                      ),
                    ],
                  ),
                ),
                SizedBox(
                  height: 17,
                ),
              ],
            ),
          ),
        ),
        Visibility(
          visible: showBottom,
          child: Column(
              children: this
                  .starNames
                  .map((e) => Column(
                        mainAxisSize: MainAxisSize.min,
                        children: <Widget>[
                          ProgressBarLayout(
                            starName: e,
                            lineHeight: this.lineHeight,
                            showBorder: showProgressBarBorder,
                            value: this.values[this.starNames.indexOf(e)],
                            showPercentage: this.showPercentage,
                            starNameStyle: this.starNameStyle,
                            percentageStyle: this.percentageStyle,
                            valueColor: this.valueColor,
                            progressBarBackgroundColor:
                                this.progressBarBackgroundColor,
                          ),
                          SizedBox(
                            height: this.spaceBetween,
                          ),
                        ],
                      ))
                  .toList()),
        )
      ],
    ));
  }
}

class StarReviewsHorizontal extends StatelessWidget {
  final int total;
  final bool showRatingNumber;

  final List<String> starNames;
  final TextStyle starNameStyle;
  final TextStyle percentageStyle;
  final bool showPercentage;

  final Color valueColor;
  final Color progressBarBackgroundColor;
  final List<double> values;

  final double starSize;
  final Color starColor;

  final double average;

  final bool showOnlyAverage;
  final bool showProgressBarBorder;
  final TextStyle averageNumberTextStyle;
  final double spaceBetween;

  final double spaceBetweenBars;
  final double lineHeight;

  final bool alignReviewsCenter;

  StarReviewsHorizontal(
      {Key key,
      @required this.total,
      @required this.starNames,
      this.starSize = 12,
      this.spaceBetweenBars = 5,
      this.lineHeight = 7,
      this.spaceBetween = 30,
      this.alignReviewsCenter = true,
      this.showProgressBarBorder = true,
      this.showOnlyAverage = false,
      this.starColor = const Color(0xffffd900),
      this.showRatingNumber = true,
      this.showPercentage = true,
      this.starNameStyle = const TextStyle(fontSize: 12),
      this.percentageStyle = const TextStyle(fontSize: 12),
      this.averageNumberTextStyle =
          const TextStyle(fontSize: 32, fontWeight: FontWeight.bold),
      this.valueColor = const Color(0xff656565),
      this.progressBarBackgroundColor = Colors.white,
      @required this.average,
      @required this.values})
      : assert(total != null),
        super(key: key) {
    if (values == null) {
      throw ArgumentError('value cannot be empty');
    }

    if (starNames.length != values.length) {
      throw ArgumentError('startNames and values\' length different');
    }

    if (starNames == null) {
      throw ArgumentError('starNames cannot be empty');
    }
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      child: Row(
        crossAxisAlignment: this.alignReviewsCenter ? CrossAxisAlignment.center : CrossAxisAlignment.start,
        children: <Widget>[
          Column(
            children: <Widget>[
              Text(
                this.average.toString(),
                style: this.averageNumberTextStyle,
              ),
              StarDisplay(
                value: this.average,
                size: starSize,
                color: this.starColor,
                starCounts: this.starNames.length,
              ),
              SizedBox(
                height: 10,
              ),
              Text(
                '${this.total} ratings',
                style: TextStyle(fontSize: 10),
              )
            ],
          ),
          SizedBox(
            width: this.spaceBetween,
          ),
          if (!this.showOnlyAverage) ...[
            Expanded(
              child: StarReviews(
                total: this.total,
                spaceBetween: this.spaceBetweenBars,
                lineHeight: this.lineHeight,
                starNames: this.starNames,
                starSize: this.starSize,
                starColor: this.starColor,
                showRatingNumber: this.showRatingNumber,
                showHeader: false,
                showBottom: true,
                showPercentage: this.showPercentage,
                starNameStyle: this.starNameStyle,
                percentageStyle: this.percentageStyle,
                valueColor: this.valueColor,
                showProgressBarBorder: this.showProgressBarBorder,
                progressBarBackgroundColor: this.progressBarBackgroundColor,
                average: this.average,
                values: this.values,
              ),
            )
          ]
        ],
      ),
    );
  }
}


lib/star_display.dart
import 'package:flutter/material.dart';

class StarDisplay extends StatelessWidget {
  final double value;
  final double size;
  final Color color;
  final int starCounts;

  const StarDisplay(
      {Key key,
        this.starCounts = 5,
        this.value = 0,
        this.size = 16,
        this.color = const Color(0xffffd900)})
      : assert(value != null),
        super(key: key);

  @override
  Widget build(BuildContext context) {
    return Row(
      mainAxisSize: MainAxisSize.min,
      children: List.generate(starCounts, (index) {
        return Icon(
          (value - index >= 0.8)
              ? Icons.star
              : (value - index >= 0.3)
              ? Icons.star_half
              : Icons.star_border,
          color: this.color,
          size: this.size,
        );
      }),
    );
  }
}

lib/progress-bar-layout.dart
import 'package:flutter/material.dart';
import 'package:starreviews/progressbar.dart';

class ProgressBarLayout extends StatelessWidget {
  final String starName;
  final TextStyle starNameStyle;
  final TextStyle percentageStyle;
  final bool showPercentage;

  final Color valueColor;
  final Color progressBarBackgroundColor;
  final double value;

  ProgressBarLayout(
        {Key key,
          this.starName,
          this.showPercentage = true,
          this.starNameStyle = const TextStyle(fontSize: 12),
          this.percentageStyle = const TextStyle(fontSize: 12),
          this.valueColor = const Color(0xff656565),
          this.progressBarBackgroundColor = Colors.white,
          this.value})
    : super(key: key) {
    if (value == null) {
    throw ArgumentError('value cannot be empty');
    }

    if (starName == null) {
      throw ArgumentError('starName cannot be empty');
    }
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      child: Row(
        children: <Widget>[
          Text(this.starName, style: this.starNameStyle),
          SizedBox(
            width: 10,
          ),
          Expanded(
            child: ProgressBar(
              value: this.value,
              valueColor: this.valueColor,
              backgroundColor: this.progressBarBackgroundColor,
            ),
          ),
          Visibility(
            visible: this.showPercentage,
            child: Row(
              children: <Widget>[
                SizedBox(
                  width: 10,
                ),
                Text((this.value * 100).toInt().toString() + '%',
                    style: this.percentageStyle)
              ],
            ),
          ),
        ],
      ),
    );
  }
}

lib/progress_bar/progress_bar.dart
import 'package:flutter/material.dart';
import 'package:percent_indicator/linear_percent_indicator.dart';

class ProgressBar extends StatelessWidget {
  final Color valueColor;
  final Color backgroundColor;
  final double value;

  final double lineHeight;

  final bool showBorder;

  ProgressBar({Key key, this.valueColor, this.lineHeight = 5, this.showBorder = true, this.backgroundColor, this.value})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Container(
      padding: EdgeInsets.all(this.showBorder ? 2 : 0),
      decoration: this.showBorder ? BoxDecoration(
          borderRadius: BorderRadius.all(Radius.circular(8)),
          border: Border.all(width: 1, color: const Color(0xffcecece))) : null,
      child: ClipRRect(
        borderRadius: BorderRadius.circular(8),
        child: Container(
          height: this.lineHeight,
          child: LinearPercentIndicator(
            lineHeight: this.lineHeight,
            padding: null,
            linearStrokeCap: LinearStrokeCap.roundAll,
            percent: this.value,
            // percent filled
            progressColor: this.valueColor,
            backgroundColor: this.backgroundColor,
          ),
        ),
      ),
    );
  }
}


lib/progress_bar/progress_bar_layout.dart
import 'package:flutter/material.dart';
import 'package:starreviews/progress_bar/progress_bar.dart';

class ProgressBarLayout extends StatelessWidget {
  final String starName;
  final TextStyle starNameStyle;
  final TextStyle percentageStyle;
  final bool showPercentage;

  final Color valueColor;
  final Color progressBarBackgroundColor;
  final double value;

  final bool showBorder;
  final double lineHeight;

  ProgressBarLayout(
      {Key key,
      this.starName,
        this.lineHeight = 5,
      this.showBorder = true,
      this.showPercentage = true,
      this.starNameStyle = const TextStyle(fontSize: 12),
      this.percentageStyle = const TextStyle(fontSize: 12),
      this.valueColor = const Color(0xff656565),
      this.progressBarBackgroundColor = Colors.white,
      this.value})
      : super(key: key) {
    if (value == null) {
      throw ArgumentError('value cannot be empty');
    }

    if (starName == null) {
      throw ArgumentError('starName cannot be empty');
    }
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      child: Row(
        children: <Widget>[
          Text(this.starName, style: this.starNameStyle),
          SizedBox(
            width: 20,
          ),
          Expanded(
            child: ProgressBar(
              lineHeight: this.lineHeight,
              showBorder: this.showBorder,
              value: this.value,
              valueColor: this.valueColor,
              backgroundColor: this.progressBarBackgroundColor,
            ),
          ),
          Visibility(
            visible: this.showPercentage,
            child: Row(
              children: <Widget>[
                SizedBox(
                  width: 20,
                ),
                Text((this.value * 100).toInt().toString() + '%',
                    style: this.percentageStyle)
              ],
            ),
          ),
        ],
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:starreviews/starreviews.dart';

void main() => runApp(App());

class App extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Title',
      home: MyApp(),
    );
  }
}

class MyApp extends StatelessWidget {

List<String> _names = ['Star 5', 'Star 4', 'Star 3', 'Star 2', 'Star 1'];
List<double> _values = [0.1, 0.3, 0.4, 0.1, 0.1];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: Container(
          padding: const EdgeInsets.all(16),
          child: Column(
            children: <Widget>[
              StarReviews(
                total: 30,
                starNames: _names,
                values: _values,
                showPercentage: true,
                average: 5,
              ),
              SizedBox(height: 30,),
            StarReviewsHorizontal(
              total: 30,
              starNames: _names,
              showProgressBarBorder: false,
              valueColor: Colors.black,
              progressBarBackgroundColor: Colors.grey.withOpacity(0.4),
              values: _values,
              showPercentage: true,
              starColor: Colors.black,
              average: 4.2,
            )
          ],
          ),
        )
      ),
    );
  }
}
